src/esw/drive/odrive_control.py

```diff
@@ -0,0 +1,656 @@
+    :param _modrive: A Modrive object that abstracts away the ODrive functions.
+    :param _pair: A string that is front, middle, or back.
+    :param _speed: A Speed object that has requested left and right wheel
+        speed.
+    :param _speed_lock: A lock that prevents multiple threads from accessing
+        self._speed simultaneously.
+    :param _state: A State object that keeps track of the current state.
+    :param _state_pub: A rospy Publisher object that is used for publishing
+        state data.
+    :param _vel_pub: A rospy Publisher object that is used for publishing
+        velocity data.
+    """
+
+    start_time: float
+    _current_lim: float
+    _id: str
+    _modrive: Modrive
+    _pair: str
+    _speed: Speed
+    _speed_lock: threading.Lock
+    _state: State
+    _state_pub: rospy.Publisher
+    _vel_pub: rospy.Publisher
+
+    def __init__(self, pair: str) -> None:
+        """Initializes the components, starting with a Disconnected State.
+        :param pair: A string that is front, middle, or back"""
+        self.start_time = t.process_time()
+        self._current_lim = rospy.get_param("/odrive/config/current_lim")
+        self._speed = Speed()
+        _odrive_ids = {
+            "front": rospy.get_param("/odrive/ids/front"),
+            "middle": rospy.get_param("/odrive/ids/middle"),
+            "back": rospy.get_param("/odrive/ids/back"),
+        }
+        self._pair = pair
+        self._id = _odrive_ids[self._pair]
+        self._speed_lock = threading.Lock()
+        self._state = DisconnectedState()
+        self._state_pub = rospy.Publisher("drive_state_data", DriveStateData, queue_size=1)
+        self._vel_pub = rospy.Publisher("drive_vel_data", DriveVelData, queue_size=1)
+
+    def change_axis_speed(self, axis: Axis, speed: float) -> None:
+        """Sets the self._speed to the requested speeds. The speeds must be in
+        the range [-1.0, 1.0].
+        :param axis: An Axis object that is either Axis.LEFT or Axis.RIGHT
+        :param speed: A float that has the requested speed of the axis.
+        """
+        assert -1.0 <= speed <= 1.0, "speed must be in range[-1.0, 1.0]"
+        self._speed_lock.acquire()
+        if axis == Axis.LEFT: 
+            self._speed = Speed(speed, self._speed.right)
+        else:
+            self._speed = Speed(self._speed.left, speed)
+        self._speed_lock.release()
+
+    def get_state_string(self) -> str:
+        """Returns the state of the ODriveBridge as a short string.
+        The form is of "[insert_odrive_state]State". e.g. state is ErrorState,
+        so ros_msg.state is Error
+        :returns: A string that is the state of the ODriveBridge.
+        """
+        state = str(self._state)
+        return state[: len(state) - len("State")]
+
+    def ros_publish_data_loop(self) -> None:
+        """Publishes velocity and current data continuously."""
+        while not rospy.is_shutdown():
+            self._publish_encoder_msg()
```

Comment:
Does this really need to live in a while without any sort of delay?? Are we actually publishing this incredibly quickly or am I missing something

Reply:
This is running on a separate thread so probably ok. A delay also shouldn't affect anything other than the publish rate so it would be ok to add probably. Either way.

Reply:
Yes... we can try testing with a delay... stuff might go wrong but worth testing

Reply:
I don't see a reason why you'd need to publish more than 10Hz or so... Just cause it's in a separate thread doesn't mean it couldn't spam the network I wouldn't think

Reply:
I would try using a rospy.Rate() so we don't publish an unnecessary amount of data

Reply:
yea it worked in the past but yea i agree with u guys ill go try to fix that

Reply:
I updated the code but I still need to test it!